pkg/groups/service_test.go

```diff
@@ -1271,3 +1279,954 @@ func TestServiceRemovePermissions(t *testing.T) {
 		})
 	}
 }
+
+func TestV1Service_ListGroups(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockService := NewMockServiceInterface(ctrl)
+	mockLogger := NewMockLoggerInterface(ctrl)
+	mockTracer := NewMockTracer(ctrl)
+	mockMonitor := monitoring.NewMockMonitorInterface(ctrl)
+	mockProvider := NewMockProviderInterface(ctrl)
+
+	mockTracer.EXPECT().Start(gomock.Any(), gomock.Any()).AnyTimes().DoAndReturn(
+		func(ctx context.Context, spanName string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
+			return ctx, trace.SpanFromContext(ctx)
+		},
+	)
+	mockProvider.EXPECT().Verifier(&oidc.Config{ClientID: "mock-client-id"}).Return(oidc.NewVerifier("", nil, &oidc.Config{
+		ClientID:                   "mock-client-id",
+		SkipExpiryCheck:            true,
+		SkipIssuerCheck:            true,
+		InsecureSkipSignatureCheck: true,
+	}))
+
+	token := "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJtb2NrLXN1YmplY3QiLCJhdWQiOiJtb2NrLWNsaWVudC1pZCIsIm5hbWUiOiJKb2huIERvZSIsImlhdCI6MTUxNjIzOTAyMn0.BdspASNsnxeXnqZXZnFnkvv-ClMq0U6X1gCIUrh9V7c"
+	principal, _ := authentication.NewJWKSTokenVerifier(mockProvider, "mock-client-id", mockTracer, mockLogger, mockMonitor).VerifyAccessToken(context.TODO(), token)
+
+	type testCase struct {
+		name           string
+		setupMocks     func()
+		contextSetup   func() context.Context
+		expectedResult []string
+		expectedError  error
+	}
```

Comment:
this looks like an interesting pattern, any chance you can explain a bit (as a pr comment) the benefits of the implementation?

Reply:
yeah, basically it's like what the other tests do. Generally, a test could be divided into three steps: arrange, act, and assert. The `testCase` tries to extract the dynamic parts out of the fixed steps.